annif/backend/yake.py

```diff
@@ -0,0 +1,191 @@
+"""Annif backend using Yake keyword extraction"""
+# Enabling this optional backend may result in GPLv3 terms to cover the
+# application, because YAKE (https://github.com/LIAAD/yake) is licensed under
+# GPLv3.
+
+import yake
+import os.path
+import re
+from collections import defaultdict
+from rdflib.namespace import SKOS
+import annif.util
+from . import backend
+from annif.suggestion import SubjectSuggestion, ListSuggestionResult
+from annif.exception import ConfigurationException
+
+
+class YakeBackend(backend.AnnifBackend):
+    """Yake based backend for Annif"""
+    name = "yake"
+    needs_subject_index = False
+
+    # defaults for uninitialized instances
+    _index = None
+    _graph = None
+    INDEX_FILE = 'yake-index'
+
+    DEFAULT_PARAMETERS = {
+        'max_ngram_size': 4,
+        'deduplication_threshold': 0.9,
+        'deduplication_algo': 'levs',
+        'window_size': 1,
+        'num_keywords': 100,
+        'features': None,
+        'label_types': ['prefLabel', 'altLabel'],
+        'remove_parentheses': False
+    }
+
+    def default_params(self):
+        params = backend.AnnifBackend.DEFAULT_PARAMETERS.copy()
+        params.update(self.DEFAULT_PARAMETERS)
+        return params
+
+    @property
+    def is_trained(self):
+        return True
+
+    @property
+    def label_types(self):
+        if type(self.params['label_types']) == str:  # Label types set by user
+            label_types = [lt.strip() for lt
+                           in self.params['label_types'].split(',')]
+            self._validate_label_types(label_types)
+        else:
+            label_types = self.params['label_types']  # The defaults
+        return [getattr(SKOS, lt) for lt in label_types]
+
+    def _validate_label_types(self, label_types):
+        for lt in label_types:
+            if lt not in ('prefLabel', 'altLabel', 'hiddenLabel'):
+                raise ConfigurationException(
+                    f'invalid label type {lt}', backend_id=self.backend_id)
+
+    def initialize(self):
+        self._initialize_index()
+
+    def _initialize_index(self):
+        if self._index is None:
+            path = os.path.join(self.datadir, self.INDEX_FILE)
+            if os.path.exists(path):
+                self._index = self._load_index(path)
+                self.debug(
+                    f'Loaded index from {path} with {len(self._index)} labels')
+            else:
+                self.info('Creating index')
+                self._index = self._create_index()
+                self._save_index(path)
+                self.info(f'Created index with {len(self._index)} labels')
+
+    def _save_index(self, path):
```

Comment:
Perhaps `annif.util.atomic_save` could be used here, to ensure that the index won't be corrupted if the process terminates in the middle of writing it?